// anmatrix/base.js
var ANMatrix = (function () {
    if (typeof BigReal == 'undefined') {
        BigReal = require('dynum').BigReal;
    }
    
    /**
     * Immutable complex number type.
     */
    function Complex(_re, _im) {
        var re = _re, im = _im;
        if (typeof re == 'number') re = BigReal.fromFloat(re);
        if (typeof im == 'number') im = BigReal.fromFloat(im);
        this.re = re || BigReal.zero;
        this.im = im || BigReal.zero;
    }

    Complex.prototype.abs = function() {
        var mag = this.re.mul(this.re).add(this.im.mul(this.im));
        return mag.sqrt();
    }

    Complex.prototype.conjugate = function() {
        return new Complex(this.re, this.im.neg());
    }

    Complex.prototype.add = function(aNum) {
        if (typeof aNum == 'number') {
            return new Complex(this.re.add(BigReal.fromFloat(aNum)), this.im);
        } else {
            if (aNum instanceof BigReal) {
                return new Complex(this.re.add(aNum), this.im);
            } else {
                return new Complex(this.re.add(aNum.re), this.im.add(aNum.im));
            }
        }
    }
    
    Complex.prototype.sub = function(aNum) {
        if (typeof aNum == 'number') {
            return new Complex(this.re.sub(BigReal.fromFloat(aNum)), this.im);
        } else {
            if (aNum instanceof BigReal) {
                return new Complex(this.re.sub(aNum), this.im);
            } else {
                return new Complex(this.re.sub(aNum.re), this.im.sub(aNum.im));
            }
        }
    }

    Complex.prototype.mul = function(aNum) {
        if (typeof aNum == 'number') {
            var numObj = BigReal.fromFloat(aNum);
            return new Complex(this.re.mul(numObj), this.im.mul(numObj));
        } else {
            if (numObj instanceof BigReal) {
                return new Complex(this.re.mul(aNum), this.im.mul(aNum));
            } else {
                return new Complex(this.re.mul(aNum.re).sub(this.im.mul(aNum.im)),
                                   this.re.mul(aNum.im).add(this.im.mul(aNum.re)));
            }
        }
    }
    
    Complex.prototype.div = function(denom) {
        if (typeof denom == 'number') return this.mul(1 / denom);
        if (denom instanceof BigReal) {
            return new Complex(this.re.div(denom), this.im.div(denom));
        } else {
            return this.mul(denom.inv());
        }
    }

    Complex.prototype.inv = function() {
        var mag = this.re.mul(this.re).add(this.im.mul(this.im));
        var re = this.re.div(mag);
        var im = this.im.div(mag).neg();
        return new Complex(re, im);
    }
    
    Complex.prototype.pow = function(power) {
        if (Math.round(power) != power) throw new Error('integer powers only');
        var result = this;
        for (var i = 1; i < power; i++) {
            result = result.mul(this);
        }
        return result;
    }
    
    Complex.prototype.phase = function() {
        var mag = this.abs();
        var angle = this.im.div(mag).arcsin();
        var pi = BigReal.fromString(BigReal.piString);
        if (!this.re.sign) {
            angle = pi.sub(angle);
        }
        if (!angle.sign) {
            var twoPi = pi.mul(BigReal.fromFloat(2));
            angle = angle.add(twoPi);
        }
        return angle;
    }
    
    Complex.prototype.sqrt = function() {
        var mag = this.abs().sqrt();
        var angle = this.phase().div(BigReal.fromFloat(2));
        var re = mag.mul(angle.cos());
        var im = mag.mul(angle.sin());
        return new Complex(re, im);
    }
    
    Complex.prototype.cbrt = function() {
        var mag = this.abs().cbrt();
        var angle = this.phase().div(BigReal.fromFloat(3));
        var re = mag.mul(angle.cos());
        var im = mag.mul(angle.sin());
        return new Complex(re, im);
    }
    
    Complex.prototype.toString = function() {
        if (!this.im.isEqual(BigReal.zero)) {
            return this.re + ' + i*' + this.im;
        }
        return this.re.toString();
    }
    
    Complex.prototype.isSmall = function() {
        var epsilonRoot = BigReal.epsilon().sqrt();
        return !this.abs().isGreater(epsilonRoot);
    }
    
    Complex.zero = new Complex(BigReal.zero, BigReal.zero);
    
    /**
     * Immutable complex valued matrix.
     */
    function Matrix(data, rows, columns) {
        this.data = data;
        this.rows = rows;
        this.cols = columns;
    }

    Matrix.zeros = function(rows, cols) {
        var data = [];
        for (var i = 0; i < rows * cols; i++) {
            data.push(new Complex(0));
        }
        return new Matrix(data, rows, cols);
    }

    Matrix.identity = function(size) {
        var mat = Matrix.zeros(size, size);
        for (var i = 0; i < size; i++) {
            mat.data[i + (i * size)] = new Complex(1);
        }
        return mat;
    }

    Matrix.prototype.copy = function() {
        var newData = [];
        for (var i = 0; i < this.data.length; i++) {
            newData[i] = this.data[i];
        }
        return new Matrix(newData, this.rows, this.cols);
    }

    Matrix.prototype.getAt = function(row, col) {
        return this.data[col + (row * this.cols)];
    }

    /**
     * Multiplies `this` on the left of aMat.
     */
    Matrix.prototype.multiply = function(aMat) {
        if (this.cols != aMat.rows) {
            throw new Error('left.cols != right.rows');
        }
    
        // multiply two complex matrices
        var rows = this.rows;
        var cols = aMat.cols;
        var data = new Array(rows * cols);
        var index = 0;
        for (var row = 0; row < rows; row++) {
            for (var col = 0; col < cols; col++) {
                data[index] = new Complex();
                for (var i = 0; i < this.cols; i++) {
                    var myValue = this.getAt(row, i);
                    var aValue = aMat.getAt(i, col);
                    data[index] = data[index].add(myValue.mul(aValue));
                }
                index++;
            }
        }
        return new Matrix(data, rows, cols);
    }

    Matrix.prototype.add = function(aMat) {
        if (aMat.rows != this.rows || aMat.cols != this.cols) {
            throw new Error('dimension mismatch');
        }
        var newData = [];
        for (var i = 0; i < this.data.length; i++) {
            newData[i] = this.data[i].add(aMat.data[i]);
        }
        return new Matrix(newData, this.rows, this.cols);
    }
    
    Matrix.prototype.scale = function(_scalar) {
        var scalar = typeof _scalar == 'number' ? new Complex(_scalar) : _scalar;
        var newData = [];
        for (var i = 0; i < this.data.length; i++) {
            newData[i] = this.data[i].mul(scalar);
        }
        return new Matrix(newData, this.rows, this.cols);
    }
    
    Matrix.prototype.transpose = function() {
        var newMat = Matrix.zeros(this.cols, this.rows);
        for (var row = 0; row < this.rows; row++) {
            for (var col = 0; col < this.cols; col++) {
                newMat.setAt(col, row, this.getAt(row, col).conjugate());
            }
        }
        return newMat;
    }
    
    Matrix.prototype.subMatrix = function(row, col, rows, cols) {
        if (row + rows > this.rows) throw new Error('rows out of bounds');
        if (col + cols > this.cols) throw new Error('cols out of bounds');
        var newMat = Matrix.zeros(rows, cols);
        for (var i = 0; i < rows; i++) {
            for (var j = 0; j < cols; j++) {
                newMat.setAt(i, j, this.getAt(i + row, j + col));
            }
        }
        return newMat;
    }
    
    Matrix.prototype.copyInto = function(aMat, row, col) {
        if (row + aMat.rows > this.rows) throw new Error('row out of bounds');
        if (col + aMat.cols > this.cols) throw new Error('col out of bounds');
        var result = this.copy();
        for (var i = 0; i < aMat.rows; i++) {
            for (var j = 0; j < aMat.cols; j++) {
                result.setAt(i + row, j + col, aMat.getAt(i, j));
            }
        }
        return result;
    }
    
    Matrix.prototype.toString = function() {
        var str = '[';
        for (var row = 0; row < this.rows; row++) {
            if (row != 0) str += '; ';
            for (var col = 0; col < this.cols; col++) {
                if (col != 0) str += ' ';
                str += this.getAt(row, col).toString();
            }
        }
        return str + ']';
    }
    
    Matrix.prototype.avg = function() {
        var sum = Complex.zero;
        for (var i = 0; i < this.data.length; i++) {
            sum = sum.add(this.data[i]);
        }
        return sum.div(this.data.length);
    }
    
    /**
     * Private: do not call this!
     */
    Matrix.prototype.setAt = function(row, col, val) {
        this.data[col + (row * this.cols)] = val;
    }
    
    // return the exports
    return {Matrix: Matrix, Complex: Complex};
})();

if (module) {
    module.exports = ANMatrix;
} else {
    window.ANMatrix = ANMatrix;
}
